Documentation: Adding technical overview (#210)
corona-warn-app/cwa-server: docs/architecture-overview.md

# Architecture CWA Backend

This document outlines the CWA back end architecture on high level. This document
does not necessarily reflect the current implementation status in this repository, as there are
still frequent changes to the exposure notification API by Google/Apple.

## Overview

The Corona Warn App ("CWA") server runs in a Kubernetes cluster on the Open Shift (“OSP”) platform.
Its main objective is to enable users to take part in the exposure notification framework based
on Apple/Google specifications. Although CWA aims at compliance to the spec on a protocol level, it

> does not mean automatically that all features will be implemented. Main driver for these assessments is data privacy and protection concerns.

Reviewer: ```suggestion
does not mean automatically that all features will be implemented. Main driver for these assessments is data privacy and protection (DPP) concerns.
```

Find the latest specifications of Google/Apple here:
- [Google Spec](https://static.googleusercontent.com/media/www.google.com/en//covid19/exposurenotifications/pdfs/Exposure-Key-File-Format-and-Verification.pdf)
- [Apple Spec](https://developer.apple.com/documentation/exposurenotification/setting_up_an_exposure_notification_server?changes=latest_beta)

On a high level, the application consists of two main parts, as shown below.

![Overview Diagram](./images/v4.png)

1. CWA Backend: Handles submission & aggregation/distribution of diagnosis keys and configuration files.

> 2. Verification Backend: Deals with test result verification and issues TANs. This back end is managed & deployed

Reviewer: consistency of spelling of "backend"

This document outlines the CWA back end components, which are part of this repository. For the full architectural
overview, check out the [solution architecture](https://github.com/corona-warn-app/cwa-documentation/blob/master/solution_architecture.md).

## Integration with Other Systems

### Object Store

All mobile app relevant files will be stored on the S3. Those files are:
-	aggregated files containing the diagnostic keys, which were reported in a specific interval (e.g. hourly)
-	Daily aggregated files containing the exposure keys, which were reported for this day.
-	Configuration files, containing the [exposure configuration](https://developer.apple.com/documentation/exposurenotification/enexposureconfiguration) and CWA mobile app config
-	Additional files regarding meta information of available files/structures/etc.

The files will be pushed to an S3 compliant object store whenever new files become available. File structure definitions for those files can be found in the respective Services chapter.

The mobile application will use a CDN for fetching files, which mirrors all files as a transparent proxy present in the object store.

### Verification Backend

> Note: The verification back end design is not yet available as discussions are ongoing.

Reviewer: Will be available soon
```suggestion
Note: The verification backend design is not yet available as discussions are ongoing.
```

> The verification back end supports the user's journey beginning at scanning the QR code printed

Reviewer: ```suggestion
The verification backend supports the user's journey beginning at scanning the QR code printed
```

> The verification back end supports the users' journey beginning at a them scanning their QR code printed

Reviewer 2: There seems to be some wording mistakes here

Reviewer 3: Thank you for pointing this out

on the COVID-19 test strip until the upload of diagnosis keys when the user was tested positive. Testing
labs will update the Verification Backend when a COVID-19 test result is positive. Since the QR code
is linked to a test, the mobile application will be able to fetch a test-positive notification from
the Verification Backend, along with a TAN. The TAN will be used as an authorization token when
the user uploads the diagnosis keys of the past 14 days.

Therefore, from a CWA Backend perspective, the Verification Backend provides an endpoint for TAN verification.

## Security

### Endpoint Protection

> The CWA back end exposes only one endpoint – the submission endpoint.

Reviewer: ```suggestion
The CWA backend exposes only one endpoint – the submission endpoint.
```

The endpoint is public (unauthenticated), and authorization for calls is granted to users who are passing a valid TAN.

> The TAN verification cannot be done on CWA back end, but the task is delegated to the Verification Backend (see Verification Backend chapter in Integration with other Systems).

Reviewer: ```suggestion
The TAN verification cannot be done on the CWA backend, but the task is delegated to the Verification Backend (see Verification Backend chapter in Integration with other Systems).
```

### Authenticity

> All by CWA published files will be digitally signed by CWA.

Reviewer: ```suggestion
All files published by CWA will be digitally signed by CWA.
```

This ensures that clients can trust the file they have received from the third-party CDN.

The proto files from Google/Apple already specify how signing should work. Each diagnosis key aggregate file
will be a zip file, containing two files - one carries the actual payload. The other one carries signature information.

CWA will be required to share the public key with Google/Apple, so that the API on the mobile client
will be able verify against it.

### Fake Submissions

In order to protect the privacy of the users, the mobile app needs to send fake submissions from time to time.
The server accepts the incoming calls and will treat them the same way as regular submissions.

> The payload & behavior of fake and real requests must be similar, so that 3rd parties are unable to differentiate between those requests.

Reviewer: ```suggestion
The payload and behavior of fake and real requests must be similar, so that 3rd parties are unable to differentiate between those requests.
```

If a submission request marked as fake is received by the server, the caller will be presented with a successful result.

> The CWA back end will not persist the entry on the database and will ensure that fake and real requests are taking the same amount of total response time, e.g. by delaying the response to the client, if necessary.

Reviewer: ```suggestion
The CWA backend will not persist the entry on the database and will ensure that fake and real request take the same amount of total response time, e.g. by delaying the response to the client, if necessary.
```

## Services

### Submission Service

> The submission services only task is to process uploaded diagnosis keys and persist them to the database after the TAN has been verified.

Reviewer: ```suggestion
The submission service's only task is to process uploaded diagnosis keys and persist them to the database after the TAN has been verified.
```

> The actual task of the verification is handed over to the verification back end, which will provide the verification result back to CWA.

Reviewer: ```suggestion
The actual task of the verification is handed over to the verification backend, which will provide the verification result back to CWA.
```

After verification was successfully done, the diagnosis keys are persisted in the database, and will be published in the next batch.

The payload send by the mobile is defined as:

    message SubmissionPayload {
      repeated Key keys = 1;
    }

    message Key {
      bytes keyData = 1; // Key of infected user
      uint32 rollingStartNumber = 2; // Interval number when the key's EKRollingPeriod started.
      uint32 rollingPeriod = 3; // Number of 10-minute windows between key rolling.
      int32 transmissionRiskLevel = 4; // Risk of transmission associated with the person this key came from.

>     }

Reviewer: use code markdown

Additionally, the endpoint requires the following headers to be set:

    CWA-Authorization: TAN <TAN>
    CWA-Fake: <0 or 1>

> There is currently no official spec for publishing diagnosis keys to the server.

Reviewer: ```suggestion
There is currently no official specification for publishing diagnosis keys to the server.
```

Google currently uses the following in their reference implementation.

[exposure_model.go](https://github.com/google/exposure-notifications-server/blob/master/internal/database/exposure_model.go)

    type Publish struct {
        Keys                      []ExposureKey `json:"temporaryExposureKeys"`
        Regions                   []string      `json:"regions"`
        AppPackageName            string        `json:"appPackageName"`
        Platform                  string        `json:"platform"`
        DeviceVerificationPayload string        `json:"deviceVerificationPayload"`
        VerificationPayload       string        `json:"verificationPayload"`
        Padding                   string        `json:"padding"`

>     }

Reviewer: code markdown

As of this moment, due to DDP concerns it is unclear whether the device attestation-related fields are in scope of CWA.

### Distribution Service

> The distribution services objective is to publish all CWA-related files to the object store, from which

Reviewer: ```suggestion
The distribution service's objective is to publish all CWA-related files to the object store, from which
```

the clients will fetch their data from. There are three types of files.

#### Key Export

Key Export files are files, which hold published diagnosis keys from COVID-19 positive tested users.
These files are based on the specification of Google/Apple and are generated in regular intervals.
Each interval generates a `.zip` file, containing two files:

1. export.bin: Contains the list of diagnosis keys
2. export.sig: Contains signature information needed for validating the export.bin file.
The file structure definition can be found [here](https://github.com/google/exposure-notifications-server/blob/master/internal/pb/export/export.proto).

The distribution service is triggered by a CRON scheduler, currently set to 1 hour. However, this
will change, since the Google/Apple API will have rate-limiting in place for a maximum of 15 checks
per day.

> 

Reviewer: ```suggestion
```
